### PR TITLE
Update checkout action to use v3 and upgrade to node16. Ref #244

### PR DESCRIPTION
**Pull request by soleo**
https://github.blog/changelog/2022-09-22-github-actions-all-actions-will-begin-running-on-node16-instead-of-node12/

GitHub is sunseting nodeJS 12 for actions. 

Similar approach like https://github.com/actions/checkout/pull/689/files

### CODE REVIEW

**Comment by bkendall** [comment]
This looks very reasonable. Would you mind catching it up with `main` and updating the branch? Thanks!

**Comment by soleo** [comment]
Sure, @bkendall . I just rebased it with the main branch. If you wanna see a working demo, you could also check out https://github.com/soleo/xinjiangshao.com/blob/master/.github/workflows/firebase-hosting-pull-request.yml for reference

File: package.json [hunk 1]
```diff
@@ -10,7 +10,7 @@
     "@babel/core": "^7.12.3",
     "@babel/preset-env": "^7.12.1",
     "@babel/preset-typescript": "^7.12.1",
-    "@tsconfig/node12": "^1.0.7",
+    "@tsconfig/node16": "^1.0.3",
```
Comment (bkendall) [inline package.json:13]: I'm 99% sure that this package isn't even required. It's not imported into the TS config, and it's not referenced _anywhere_. Want to just remove it? :)

**Comment by bkendall** [comment]
oh this nonsense. The tests that require secrets are going to fail because they won't work from a fork. I'll take care of that once you remove the tsconfig thing :)

File: package.json (hunk 1 above)
Comment (soleo) [reply package.json:13]: ohh, cool. I will remove it

Comment (soleo) [reply package.json]: @bkendall now removed `@tsconfig/node16` and I squashed my commits. Ready for another review if you are free.

**Review by bc-micah — APPROVED** [review]

**Pull request closed without merging**